quipucords/api/credential/serializer.py

```diff
@@ -23,225 +22,180 @@ def expand_filepath(filepath):
+    # [2]: https://github.com/encode/django-rest-framework/blob/3.14.0/rest_framework/serializers.py#L121  # noqa: E501
+
+    def __new__(
+        cls, instance=None, data=empty, **kwargs
+    ):  # pylint: disable=arguments-differ
+        """Overloaded __new__ to return the adequate serializer."""
+        cred_type = cls._get_cred_type(instance, data)
+        klass = cls._get_serializer_class(cred_type)
+        return super().__new__(klass, instance=instance, data=data, **kwargs)
+
+    @classmethod
+    def _get_serializer_class(cls, cred_type):
+        """Return the adequate serializer based on 'cred_type'."""
+
+        def _default():
+            return CredentialSerializer
+
+        serializer_per_datasource = defaultdict(
+            _default,
+            {
+                DataSources.NETWORK: NetworkCredentialSerializer,
+                DataSources.OPENSHIFT: AuthTokenSerializer,
+                DataSources.VCENTER: UsernamePasswordSerializer,
+                DataSources.SATELLITE: UsernamePasswordSerializer,
+            },
         )
 
-        if "cred_type" not in validated_data:
-            error = {"cred_type": [_(messages.CRED_TYPE_REQUIRED_CREATED)]}
-            raise ValidationError(error)
-
-        cred_type = validated_data.get("cred_type")
-        become_method = validated_data.get("become_method")
-        become_user = validated_data.get("become_user")
-
-        if cred_type == DataSources.NETWORK and not become_method:
-            # Set the default become_method to be sudo if not specified
-            validated_data["become_method"] = Credential.BECOME_SUDO
-        if cred_type == DataSources.NETWORK and not become_user:
-            # Set the default become_user to root if not specified
-            validated_data["become_user"] = Credential.BECOME_USER_DEFAULT
-
-        return super().create(validated_data)
-
-    def update(self, instance, validated_data):
-        """Update a host credential."""
-        name = validated_data.get("name")
-        check_for_existing_name(
-            Credential.objects,
-            name,
-            _(messages.HC_NAME_ALREADY_EXISTS % name),
-            search_id=instance.id,
-        )
+        return serializer_per_datasource[cred_type]
+
+    @classmethod
+    def _get_cred_type(cls, instance, data):
+        if isinstance(instance, Credential):
+            return instance.cred_type
+        if isinstance(data, dict):
+            return data.get("cred_type")
+        return None
+
+    def __init__(self, instance=None, data=empty, **kwargs):
+        """Customize class initialization."""
+        if instance and isinstance(data, dict):
+            # assume cred_type == instance.cred_type if not provided
+            # this is only required to avoid breaking current functionality
+            # that was treating 'cred_type' as optional for updates and required for
+            # credential creation
```

Comment:
Is this referring to a behavior of external clients like the web UI or qpc CLI that would break? Or is this something internal to quipucords that we could refactor independent of the clients, and if so, we could remove this condition?

Reply:
referring to UI and maybe CLI